### PR TITLE
Support 3.12

### PR DESCRIPTION
**Pull request by JelleZijlstra**

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
I guess this still runs alpha 7, which doesn't have a lot of stuff we need. Let's try again after beta 1 is released.

**Comment by AlexWaygood** [comment]
We could use https://github.com/deadsnakes/action to test with a nightly build in CI

**Comment by AlexWaygood** [comment]
Guessing it'll be a few more days before `actions/setup-python` starts using 3.12b1 for `3.12-dev`

**Comment by JelleZijlstra** [comment]
What, it was released like two hours ago :D 

I think I'll cut the 4.6.0 release tonight without waiting for this, after manually confirming that things work on current 3.12.

**Comment by tmke8** [comment]
Seems 3.12-beta1 was just made available (3 hours ago): https://github.com/actions/python-versions/releases/tag/3.12.0-beta.1-5071360807

**Comment by JelleZijlstra** [comment]
Looks like we're still getting alpha7

**Comment by tmke8** [comment]
Sorry, it seems the manifest needs to be updated still:

- https://github.com/actions/python-versions/pull/228

**Comment by AlexWaygood** [comment]
> We're now getting 3.12-beta1 in CI, but the new test that was added in #188 is failing on 3.12.

It's a bug in CPython. Writing an issue up now.

**Comment by AlexWaygood** [comment]
> > We're now getting 3.12-beta1 in CI, but the new test that was added in #188 is failing on 3.12.
> 
> It's a bug in CPython. Writing an issue up now.

- #104935

**Comment by AlexWaygood** [comment]
Following https://github.com/python/typing_extensions/commit/5d51455cf9edfa06d629269274b2f128741300b4, everything should now be green on 3.12 if you merge in `main`

File: CHANGELOG.md [hunk 1]
```diff
@@ -98,7 +98,7 @@
   as a class rather than a function). This allows user-defined `NewType`s to be
   pickled. Patch by Alex Waygood.
 - Fix tests and import on Python 3.12, where `typing.TypeVar` can no longer be
-  subclassed. Patch by Jelle Zijlstra.
+  subclassed. Add explicit support for Python 3.12. Patches by Jelle Zijlstra.
```
Comment (AlexWaygood) [inline CHANGELOG.md:101]: This needs to be moved to the top of the Changelog

**Review by srittau — APPROVED** [review]

**Pull request merged**